From 2fa6531526e66521aded8943b459e3b9a1d5533e Mon Sep 17 00:00:00 2001
From: tonishi92 <tatsuo.onishi@gmail.com>
Date: Tue, 13 Dec 2022 17:06:55 +0100
Subject: [PATCH 66/70] (Patch) Undo a previous commit on irr_rates (121aa95).

Description:
  To apply a next patch on irr_rates, a previous commit on irr_rates (121aa95) is undone.
  In the previous commit, a diagnostic array "irr_rates" is commented out
  because it raised an error when performed with a "backtrace" option.

  In the next commit, "irr_rates" will be maintained. In this commit, we undo the previous
  commit so that the next commit will be independent of the previous one.

List of modified files:
  * chem/KPP/inc/mozart_mosaic_4bin/kpp_mechd_ia_mozart_mosaic_4bin.inc
  * chem/KPP/inc/mozart_mosaic_4bin_aq/kpp_mechd_ia_mozart_mosaic_4bin_aq.inc
  * chem/KPP/inc/mozcart/kpp_mechd_ia_mozcart.inc
  * chem/KPP/inc/t1_mozcart/kpp_mechd_ia_t1_mozcart.inc
  * chem/chem_driver.F
---
 .../kpp_mechd_ia_mozart_mosaic_4bin.inc       |  12 +-
 .../kpp_mechd_ia_mozart_mosaic_4bin_aq.inc    |  12 +-
 chem/KPP/inc/mozcart/kpp_mechd_ia_mozcart.inc |  12 +-
 .../t1_mozcart/kpp_mechd_ia_t1_mozcart.inc    |  12 +-
 chem/chem_driver.F                            | 162 +++++++++---------
 5 files changed, 100 insertions(+), 110 deletions(-)

diff --git a/chem/KPP/inc/mozart_mosaic_4bin/kpp_mechd_ia_mozart_mosaic_4bin.inc b/chem/KPP/inc/mozart_mosaic_4bin/kpp_mechd_ia_mozart_mosaic_4bin.inc
index b2df013..1e4742f 100644
--- a/chem/KPP/inc/mozart_mosaic_4bin/kpp_mechd_ia_mozart_mosaic_4bin.inc
+++ b/chem/KPP/inc/mozart_mosaic_4bin/kpp_mechd_ia_mozart_mosaic_4bin.inc
@@ -1,8 +1,6 @@
 !
-!TO2   IF( irr_option(dm) ) THEN
-!        DO n = param_first_scalar,irr_diag_cnt(dm)
-!          irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
-!        ENDDO
-!      ENDIF
-!   
-!TO2
+   IF( irr_option(dm) ) THEN
+     DO n = param_first_scalar,irr_diag_cnt(dm)
+       irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
+     ENDDO
+   ENDIF
diff --git a/chem/KPP/inc/mozart_mosaic_4bin_aq/kpp_mechd_ia_mozart_mosaic_4bin_aq.inc b/chem/KPP/inc/mozart_mosaic_4bin_aq/kpp_mechd_ia_mozart_mosaic_4bin_aq.inc
index c564438..1cf014c 100644
--- a/chem/KPP/inc/mozart_mosaic_4bin_aq/kpp_mechd_ia_mozart_mosaic_4bin_aq.inc
+++ b/chem/KPP/inc/mozart_mosaic_4bin_aq/kpp_mechd_ia_mozart_mosaic_4bin_aq.inc
@@ -1,8 +1,6 @@
 !
-!TO2   IF( irr_option(dm) ) THEN
-!        DO n = param_first_scalar,num_irr_diag
-!          irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
-!        ENDDO
-!      ENDIF
-!   
-!TO2
+   IF( irr_option(dm) ) THEN
+     DO n = param_first_scalar,num_irr_diag
+       irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
+     ENDDO
+   ENDIF
diff --git a/chem/KPP/inc/mozcart/kpp_mechd_ia_mozcart.inc b/chem/KPP/inc/mozcart/kpp_mechd_ia_mozcart.inc
index c564438..1cf014c 100644
--- a/chem/KPP/inc/mozcart/kpp_mechd_ia_mozcart.inc
+++ b/chem/KPP/inc/mozcart/kpp_mechd_ia_mozcart.inc
@@ -1,8 +1,6 @@
 !
-!TO2   IF( irr_option(dm) ) THEN
-!        DO n = param_first_scalar,num_irr_diag
-!          irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
-!        ENDDO
-!      ENDIF
-!   
-!TO2
+   IF( irr_option(dm) ) THEN
+     DO n = param_first_scalar,num_irr_diag
+       irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
+     ENDDO
+   ENDIF
diff --git a/chem/KPP/inc/t1_mozcart/kpp_mechd_ia_t1_mozcart.inc b/chem/KPP/inc/t1_mozcart/kpp_mechd_ia_t1_mozcart.inc
index c564438..1cf014c 100644
--- a/chem/KPP/inc/t1_mozcart/kpp_mechd_ia_t1_mozcart.inc
+++ b/chem/KPP/inc/t1_mozcart/kpp_mechd_ia_t1_mozcart.inc
@@ -1,8 +1,6 @@
 !
-!TO2   IF( irr_option(dm) ) THEN
-!        DO n = param_first_scalar,num_irr_diag
-!          irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
-!        ENDDO
-!      ENDIF
-!   
-!TO2
+   IF( irr_option(dm) ) THEN
+     DO n = param_first_scalar,num_irr_diag
+       irr_rates(i,k,j,n) = real( IRR_WRK(irr_diag_ndx(n-1,dm)),kind=4 )
+     ENDDO
+   ENDIF
diff --git a/chem/chem_driver.F b/chem/chem_driver.F
index 8e99e46..834bb52 100755
--- a/chem/chem_driver.F
+++ b/chem/chem_driver.F
@@ -300,8 +300,8 @@
       INTEGER                         :: ksubt
 
       REAL :: chem_minval, dtstepc
-! TO2
-!      REAL, DIMENSION(:,:,:,:), ALLOCATABLE :: irr_rates
+
+      REAL, DIMENSION(:,:,:,:), ALLOCATABLE :: irr_rates
 
       REAL, DIMENSION(grid%sm31:grid%em31, grid%sm32:grid%em32, grid%sm33:grid%em33, numgas_mam) :: gas_aqfrac ! fraction of gas that is in cloud water
       
@@ -575,27 +575,27 @@
      CASE (MOZART_MOSAIC_4BIN_KPP) 
        CALL wrf_debug(15,'calling mozart_mosaic_4bin_kpp from chem_driver')
 #if( WRF_CHEM == 1 && WRF_KPP == 1 )
-!TO2       IF( grid%irr_opt == 1 ) then
-!TO2         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_mozart_mosaic_4bin),stat=astat )
-!TO2         IF( astat /= 0 ) THEN
-!TO2           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
-!TO2           CALL wrf_error_fatal( trim(msg) )
-!TO2         ENDIF
-!TO2         num_irr_diag = num_irr_diag_mozart_mosaic_4bin
-!TO2       ENDIF
+       IF( grid%irr_opt == 1 ) then
+         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_mozart_mosaic_4bin),stat=astat )
+         IF( astat /= 0 ) THEN
+           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
+           CALL wrf_error_fatal( trim(msg) )
+         ENDIF
+         num_irr_diag = num_irr_diag_mozart_mosaic_4bin
+       ENDIF
 #endif
        haveaer = .true.
      CASE (MOZART_MOSAIC_4BIN_AQ_KPP)
        CALL wrf_debug(15,'calling mozart_mosaic_4bin_aq_kpp from chem_driver')
 #if( WRF_CHEM == 1 && WRF_KPP == 1 )
-!TO2       IF( grid%irr_opt == 1 ) then
-!TO2         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_mozart_mosaic_4bin_aq),stat=astat )
-!TO2         IF( astat /= 0 ) THEN
-!TO2           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
-!TO2           CALL wrf_error_fatal( trim(msg) )
-!TO2         ENDIF
-!TO2         num_irr_diag = num_irr_diag_mozart_mosaic_4bin_aq
-!TO2       ENDIF
+       IF( grid%irr_opt == 1 ) then
+         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_mozart_mosaic_4bin_aq),stat=astat )
+         IF( astat /= 0 ) THEN
+           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
+           CALL wrf_error_fatal( trim(msg) )
+         ENDIF
+         num_irr_diag = num_irr_diag_mozart_mosaic_4bin_aq
+       ENDIF
 #endif
        haveaer = .true.
        !BSINGH -  Added case statement for SAPRC 8 bin
@@ -659,26 +659,26 @@
      CASE (MOZCART_KPP)
        CALL wrf_debug(15,'calling mozcart driver from chem_driver')
 #if( WRF_CHEM == 1 && WRF_KPP == 1 )
-!TO2       IF( grid%irr_opt == 1 ) then
-!TO2         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_mozcart),stat=astat )
-!TO2         IF( astat /= 0 ) THEN
-!TO2           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
-!TO2           CALL wrf_error_fatal( trim(msg) )
-!TO2         ENDIF
-!TO2         num_irr_diag = num_irr_diag_mozcart
-!TO2       ENDIF
+       IF( grid%irr_opt == 1 ) then
+         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_mozcart),stat=astat )
+         IF( astat /= 0 ) THEN
+           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
+           CALL wrf_error_fatal( trim(msg) )
+         ENDIF
+         num_irr_diag = num_irr_diag_mozcart
+       ENDIF
 #endif
      CASE (T1_MOZCART_KPP)
        CALL wrf_debug(15,'calling t1_mozcart driver from chem_driver')
 #if( WRF_CHEM == 1 && WRF_KPP == 1 )
-!TO2       IF( grid%irr_opt == 1 ) then
-!TO2         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_t1_mozcart),stat=astat )
-!TO2         IF( astat /= 0 ) THEN
-!TO2           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
-!TO2           CALL wrf_error_fatal( trim(msg) )
-!TO2         ENDIF
-!TO2         num_irr_diag = num_irr_diag_t1_mozcart
-!TO2       ENDIF
+       IF( grid%irr_opt == 1 ) then
+         ALLOCATE( irr_rates(grid%sm31:grid%em31,grid%sm32:grid%em32,grid%sm33:grid%em33,num_irr_diag_t1_mozcart),stat=astat )
+         IF( astat /= 0 ) THEN
+           write(msg,'(''chem_driver: Failed to allocate irr_rates; error = '',i8)') astat
+           CALL wrf_error_fatal( trim(msg) )
+         ENDIF
+         num_irr_diag = num_irr_diag_t1_mozcart
+       ENDIF
 #endif
      CASE (CB05_SORG_AQ_KPP)
        CALL wrf_debug(15,'calling cb05_sorg_aq_kpp from chem_driver')
@@ -1513,56 +1513,54 @@ CALL kpp_mechanism_driver (chem,
 !
    ids,ide, jds,jde, kds,kde,                                                         &
    ims,ime, jms,jme, kms,kme,                                                         &
-!TO2
-!   its,ite,jts,jte,kts,kte,grid%id,num_irr_diag,irr_rates)
-   its,ite,jts,jte,kts,kte,grid%id,num_irr_diag)
+   its,ite,jts,jte,kts,kte,grid%id,num_irr_diag,irr_rates)
           if( chm_is_mozart ) then
              call mozcart_lbc_set( chem, num_chem, grid%id, &
                                    ims, ime, jms, jme, kms, kme,    &
                                     its, ite, jts, jte, kts )
           end if
-    !TO2 IF( grid%irr_opt == 1 ) then
-    !TO2   select case( config_flags%chem_opt )
-    !TO2     case( mozcart_kpp )
-    !TO2       do n=param_first_scalar,num_irr_diag
-    !TO2         do j=jts,jte
-    !TO2           do k=kts,kte
-    !TO2             irr_diag_mozcart(its:ite,k,j,n) = irr_diag_mozcart(its:ite,k,j,n) &
-    !TO2              + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
-    !TO2           enddo
-    !TO2         enddo
-    !TO2       enddo
-    !TO2     case( t1_mozcart_kpp )
-    !TO2       do n=param_first_scalar,num_irr_diag
-    !TO2         do j=jts,jte
-    !TO2           do k=kts,kte
-    !TO2             irr_diag_t1_mozcart(its:ite,k,j,n) = irr_diag_t1_mozcart(its:ite,k,j,n) &
-    !TO2              + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
-    !TO2           enddo
-    !TO2         enddo
-    !TO2       enddo
-    !TO2     case( mozart_mosaic_4bin_kpp )
-    !TO2       do n=param_first_scalar,num_irr_diag
-    !TO2         do j=jts,jte
-    !TO2           do k=kts,kte
-    !TO2             irr_diag_mozart_mosaic_4bin(its:ite,k,j,n) = &
-    !TO2                   irr_diag_mozart_mosaic_4bin(its:ite,k,j,n) &
-    !TO2                   + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
-    !TO2           enddo
-    !TO2         enddo
-    !TO2       enddo
-    !TO2     case( mozart_mosaic_4bin_aq_kpp )
-    !TO2       do n=param_first_scalar,num_irr_diag
-    !TO2         do j=jts,jte
-    !TO2           do k=kts,kte
-    !TO2             irr_diag_mozart_mosaic_4bin_aq(its:ite,k,j,n) = &
-    !TO2                   irr_diag_mozart_mosaic_4bin_aq(its:ite,k,j,n) &
-    !TO2                   + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
-    !TO2           enddo
-    !TO2         enddo
-    !TO2       enddo
-    !TO2   end select
-    !TO2 ENDIF
+    IF( grid%irr_opt == 1 ) then
+      select case( config_flags%chem_opt )
+        case( mozcart_kpp )
+          do n=param_first_scalar,num_irr_diag
+            do j=jts,jte
+              do k=kts,kte
+                irr_diag_mozcart(its:ite,k,j,n) = irr_diag_mozcart(its:ite,k,j,n) &
+                 + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
+              enddo
+            enddo
+          enddo
+        case( t1_mozcart_kpp )
+          do n=param_first_scalar,num_irr_diag
+            do j=jts,jte
+              do k=kts,kte
+                irr_diag_t1_mozcart(its:ite,k,j,n) = irr_diag_t1_mozcart(its:ite,k,j,n) &
+                 + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
+              enddo
+            enddo
+          enddo
+        case( mozart_mosaic_4bin_kpp )
+          do n=param_first_scalar,num_irr_diag
+            do j=jts,jte
+              do k=kts,kte
+                irr_diag_mozart_mosaic_4bin(its:ite,k,j,n) = &
+                      irr_diag_mozart_mosaic_4bin(its:ite,k,j,n) &
+                      + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
+              enddo
+            enddo
+          enddo
+        case( mozart_mosaic_4bin_aq_kpp )
+          do n=param_first_scalar,num_irr_diag
+            do j=jts,jte
+              do k=kts,kte
+                irr_diag_mozart_mosaic_4bin_aq(its:ite,k,j,n) = &
+                      irr_diag_mozart_mosaic_4bin_aq(its:ite,k,j,n) &
+                      + dtstepc*irr_rates(its:ite,k,j,n)*1.e6/(dens2con_a*rho(its:ite,k,j))
+              enddo
+            enddo
+          enddo
+      end select
+    ENDIF
    
     if(config_flags%chem_opt == 301 ) then
        chem(its:ite,kts:kte,jts:jte,p_sulf)=vcsulf_old(its:ite,kts:kte,jts:jte)
@@ -2058,9 +2056,9 @@ end if !Chemistry time step check
    grid%dgnumwet_a2(its:ite, kts:kte, jts:jte) = grid%dgnumwet4d(its:ite, kts:kte, jts:jte, 2)
    grid%dgnumwet_a3(its:ite, kts:kte, jts:jte) = grid%dgnumwet4d(its:ite, kts:kte, jts:jte, 3)
 
-   !TO2 IF( allocated( irr_rates ) ) THEN
-   !TO2   deallocate( irr_rates )
-   !TO2 ENDIF
+   IF( allocated( irr_rates ) ) THEN
+     deallocate( irr_rates )
+   ENDIF
 
 
     END subroutine chem_driver
-- 
2.31.1

